### PR TITLE
Settings window

### PR DESCRIPTION
**Pull request by DanVanAtta**
Based on #844

Adds a settings window to the game. Can be accessed from the engine preferences screen or from the 'game' menu when in-game.

New packages/classes:
settings_package

Scroll settings tab:
__scrolling_settings

Folders settings tab:
__folders

__battle_calc_settings

AI options tab:
__ai_pause

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/846)

### CODE REVIEW

**Comment by RoiEXLab** [comment]
> Based on #644
You mean #844...

**Comment by RoiEXLab** [comment]
@DanVanAtta small suggestion:
As a user you might ask yourself what unit of measurement these values are...
E.g. the Pause duration should be seconds, millliseconds or whatever (suggesting seconds, since they are more common for non-developers)

**Comment by DanVanAtta** [comment]
Fixed the 644-> 844 in the description.

* The description text is still a bit rough. Making sure there are unit labels is a good suggestion. The valid ranges need to be revisited a bit too, some allow much too high of values.

* For the moment the settings window can only handle text input values. I think I can update it so we can have arbitrary swing components for the input entry, which would allow radio buttons and/or check boxes. Though, a text only input is still quite useful for the various numeric valued settings.

**Comment by Cernelius** [comment]
Nice addition, except the ability to choose the map's folder, that it may be handy, but I think it should not be, to save yourselves issues about issues, etc..

Also, you should call the new map folder something like "maps-1_9", not "downloadedMaps", as this would be confusing to regular users (especially with reference to the "Open User Maps and Savegames Folder" option), and would give prejudice to changing it yet again.

Some related notes:

- I've always told Veqryn that the default for dice calculations should be 1,000 (stead 2,000) and the maximum (for both dice and LL) should be 100,000 (stead 20,000), but he always rebuked, so I'm telling you now.

- I still believe the AI Pause Duration default should be 250 ms.

- I think the "Odds Calculator" or "Battle Calculator" should be renamed to "Battle Simulator", because, while it does calculations, starting from the raw data obtained from the simulations, its main point is simulating battles, not calculating the outcomes directly.

- I think it should be possible to set the percentage approssimation (default at 1% is fine, but could be 0.1%) (like having "Attacker Wins: %92.170", by setting it at 3 decimals, instead of "Attacker Wins: %92"); like an option for having decimals, from 0 to 6, with default at 0 or 1.

- I think the meta-power is dumb and I would remove it and anything related or change it to a decent formula.

**Comment by DanVanAtta** [comment]
@Cernelius interesting points, but most of those points relate to code that I have neither modified nor wrote, and is not in scope of this PR. 

I don't think I agree about the save folder setting being confusing. The 'open folder' button will open up the corresponding folder, whatever the setting. If anything, having an explicit label of  for the save folder path, which is added here, along with a description of what the setting is, if anything will cut down on confusion.

The DownloadedMaps folder rename was done some time ago. "maps-1_9" I do not like as a name since it ties maps to the game engine version. If you feel there is value in a different name, please open an issue.

**Comment by DanVanAtta** [comment]
https://github.com/triplea-game/triplea/issues/859 to discuss the topics out of scope from this PR

**Comment by RoiEXLab** [comment]
@DanVanAtta please resolve the Merge Conflicts

**Comment by DanVanAtta** [comment]
- Added a mechanism where arbitrary input Panels can be used. Added one option to use radio buttons so far:
radio_buttons

- Also updated the buttons to look more reasonable, fixed some spacing issues, and added a border around each input row (the border is not the best, but it helps split things up, perhaps a single horizontal line between rows might be better
scrolling_ui

Still to do:
- Move more of the file menu settings options to the settings window.
- Fix up setting names and descriptions
- See to removing the default battle calc count for LL, just have one default battle window calc count.
- Fix tabs, "AI" tab needs more width:
tabs
- Some Q/A around settings of radio buttons, when there is no default value we get Nulls back, can cause NPEs
- make sure all of the description areas have valid range boundaries on them

Future work:
- In places where the option is a directory, add a button to open the directory
- Consolidate all of the engine settings window with this preference window
- Improve input validation UX. Entering in invalid inputs is not the best experience right now. For example, an invalid value should revert to the current value when you click save. Alternatively and/or even better, validate the input after a user has typed, and shade the input box a red color if it is invalid.

 
I'll also see about re-organizing this PR a bit, but overall it's not a small update and is going to be a good bit of code to review.

**Comment by DanVanAtta** [comment]
Some additional updates:
- Tab width made uniform
- Invalid UI settings are reverted when user clicks save
- Improved settings description

Also re-organized the commits to make it a bit more reviewable. @ron-murhammer , @RoiEXLab  - please have a look

File: src/games/strategy/engine/ClientContext.java [hunk 1]
```diff
@@ -43,6 +46,9 @@
   private final MapListingSource mapListingSource;
   private final ScrollSettings scrollSettings;
   private final FolderSettings folderSettings;
+  private final AiSettings aiSettings;
```
Comment (DanVanAtta) [inline src/games/strategy/engine/ClientContext.java]: ClientContext wires in the settings objects. This decouples the settings window and the settings value. One day soon we will replace ClientContext with dagger. For now, the 'proper' way to get a settings value is to get it via ClientContext, with constructor dependency injection. Grant it, we are in a half-way state, so this ideal is not quite reached, but that is the goal.

**Comment by ron-murhammer** [comment]
@DanVanAtta Awesome. Definitely a step in the right direction here and really like it overall. Some comments on functionality:

1. Would be great if 'Use Defaults' instead of immediately reverting and saving default values instead set all the default values in the tab so you could see/review them then click save to accept.

2. Prefer if we swap button order and have "Save" before "Close".

3. Could definitely use some improvements to valid value ranges and descriptions but that can probably be done later on.

4. Might be good to re-order tabs so those that are used most often appear first (folders should probably be last).

I'll try to review the commits the first half of this week.

**Comment by DanVanAtta** [comment]
Good feedback. One problem with the revert behavior, I agree that would be better behavior as you describe @ron-murhammer . But - the revert button closes the settings window to hide the fact that there is no wiring to update the values of the on-screen swing components. 

It could be done, but it's not a small tweak. Currently settings are defined as a system prop getter/setter pair in a data object, and secondly as an input component that lives on a 'settings tab'. To create an input component, you define the getter/setter operation of the system property object, and a 'reader' function for a swing component. That is then used to feed back in to the system property. 

So we would need to add a 'writer' function to update the swing component, providing full symmetry. For a text field component, this would be easy. For radio buttons, we'd have to get a bit better about mapping enum values to radio buttons.

**Comment by Cernelius** [comment]
> I don't think I agree about the save folder setting being confusing. The 'open folder' button will open up the corresponding folder, corresponding to the setting value. If anything, having an explicit label for what the folder path is, which is added here, along with a description will hopefully do the opposite and cut down on confusion.

@DanVanAtta 
So, are you going to delete the current "Open User Maps and Savegame Folder" (that doesn't do what it says, but, instead, opens the "triplea" folder where both those folders are) and, instead, add two buttons like:
- "Open Downloaded Maps Folder", that will open the "downloadedMaps" folder or whatever else in the Map Download Path
- "Open Saved Games Folder", that will open the "savedGames" folder or whatever else in the Save game path

Right?

File: src/games/strategy/triplea/oddsCalculator/ta/IOddsCalculator.java [hunk 2]
```diff
@@ -52,5 +52,6 @@ AggregateResults setCalculateDataAndCalculate(final PlayerID attacker, final Pla
 
   void addOddsCalculatorListener(final OddsCalculatorListener listener);
 
+  // TODO: this method appears to never used.
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/oddsCalculator/ta/IOddsCalculator.java]: Remove it then?

File: src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java [hunk 3]
```diff
@@ -767,7 +767,11 @@ private void createComponents() {
     m_numRuns.setColumns(4);
     m_numRuns.setMin(1);
     m_numRuns.setMax(20000);
-    m_numRuns.setValue((games.strategy.triplea.Properties.getLow_Luck(m_data) ? 500 : 2000));
+
+    int simulationCount =
+        Properties.getLow_Luck(m_data) ? ClientContext.battleCalcSettings().getSimulationCountLowLuck()
+            : ClientContext.battleCalcSettings().getSimulationCountDice();
+    m_numRuns.setValue(simulationCount);
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java]: Removal of magic values :+1:

File: src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java [hunk 4]
```diff
@@ -1059,23 +1063,25 @@ private void notifyListeners() {
...
   private final UnitCategory m_category;
   private final ScrollableTextField m_textField;
+  private final GameData m_data;
   private final ListenerList<WidgetChangedListener> m_listeners = new ListenerList<>();
+  private final ScrollableTextFieldListener m_listenerTextField = new ScrollableTextFieldListener() {
+    @Override
+    public void changedValue(final ScrollableTextField field) {
+      notifyListeners();
+    }
+  };
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java]: Is lambda possible here?

File: src/games/strategy/triplea/settings/SettingInputComponent.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
+package games.strategy.triplea.settings;
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/settings/SettingInputComponent.java]: RoiEX review stop
Just a bookmark, where I stopped reviewing...

File: src/games/strategy/triplea/oddsCalculator/ta/IOddsCalculator.java (hunk 2 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/oddsCalculator/ta/IOddsCalculator.java]: Out of scope for this PR, but marking it so we don't lose the information so easily.

Comment (DanVanAtta) [inline src/games/strategy/triplea/oddsCalculator/ta/IOddsCalculator.java]: IE: such a change should be in its own PR, it is completely independent of this work.

File: src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java]: I do not think so,  ScrollableTextFieldListener not a Functional interface. We could create a wrapper to make it so, but meh..

**Comment by DanVanAtta** [comment]
@Cernelius , I think your understanding is correct, with one exception, no buttons are being removed as part of this PR. Just the new settings window is being added, and the settings it enables wired in to the rest of the game (so that way the settings window actually works). It's a future effort to merge the stuff in the 'engine preference' window with the 'settings window'

**Comment by DanVanAtta** [comment]
@Cernelius to be clear/more specific, some menu items are being removed from the 'game' menu, for example 'focus on own casualties', the settings window will be the new home for those settings. The reasoning being is that the settings window allows for the description field to be displayed next to the setting, so users will have more context as to what the setting does.

File: src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java (hunk 4 above)
Comment (RoiEXLab) [inline src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java]: `@FunctionalInterface` ?

File: src/games/strategy/triplea/settings/SettingsWindow.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+ * - save settings
+ * - close window
+ */
+public class SettingsWindow extends SwingComponents.ModalJDialog {
+
+  private static final int MAX_WIDTH = 1100;
+  private static final int TEXT_LABEL_WIDTH = MAX_WIDTH / 4;
+  private static final int ROW_HEIGHT = 60;
+
+  public static void main(String[] args) {
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/settings/SettingsWindow.java]: Please mark with a TODO for later removal...

File: src/games/strategy/triplea/settings/SystemPreferenceKey.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  FASTER_ARROW_KEY_SCROLL_MULTIPLIER,
+  WHEEL_SCROLL_AMOUNT,
+  MAP_EDGE_SCROLL_SPEED,
+  USER_MAPS_FOLDER_PATH,
+  SAVE_GAMES_FOLDER_PATH,
+  AI_PAUSE_DURACTION,
+  BATTLE_CALC_SIMULATION_COUNT_DICE,
+  BATTLE_CALC_SIMULATION_COUNT_LOW_LUCK,
+  CONFIRM_ENEMY_CASUALTIES,
+  CONFIRM_DEFENSIVE_ROLLS, FOCUS_ON_OWN_CASUALTIES, SHOW_BATTLES_BETWEEN_AI, ARROW_KEY_SCROLL_SPEED
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/settings/SystemPreferenceKey.java]: Formatting: Please move every enum in a new line...

**Comment by RoiEXLab** [comment]
@DanVanAtta looks good! IMO Ready to merge if you finished...

**Comment by RoiEXLab** [comment]
@DanVanAtta good! Nothing more to say...

**Comment by ron-murhammer** [comment]
@DanVanAtta Merging this as this looks good and is a great start at making settings more manageable and adjustable. There are a few minor things that I'll mention in case someone wants to improve on this in the future:

1. Some of the classes look like they need formatted and there are a few places with double blank lines.
2. I'd like to see 'Set Defaults' work differently by presenting values then asking to confirm.
3. Probably would also be nice to show defaults on the right description just like the range.
4. Many of the defaults and ranges could use improvement.
5. Would be nice to add in sliders for many of the ranges.
6. Would be nice to center input boxes vertically within the rows (currently closer to top)
7. Descriptions wrap in the middle of words instead of between them (ex. "How close the mouse cursor must be to the edge of the map for the map to scroll")
